src/measure/msd.cu

```diff
@@ -130,7 +130,59 @@ __global__ void gpu_find_msd(
   }
 }
 
-} // namespace
+  __global__ void gpu_find_msd_per_group(
+    const int num_atoms,
+    const int num_groups,
+    const int correlation_step,
+    const int num_correlation_steps,
+    const int* group_ids,
+    const double* g_x,
+    const double* g_y,
+    const double* g_z,
+    const double* g_x_start,
+    const double* g_y_start,
+    const double* g_z_start,
+    double* g_msd_x,  // [num_groups * time_lags]
+    double* g_msd_y,
+    double* g_msd_z)
+  {
+    // Do not use a reduction, instead handle each atom independently.
+    // Reduction does not work well, since atoms in the same group are not necessarily
+    // close in memory.
+
+    // Rewrite as a reduction over each group?
+    // Need the size_sum! Sum 0 to l, 1 to l, etc.
+    int tid = threadIdx.x;
+    int bid = blockIdx.x;
+    int size_sum = bid * num_atoms;
+    int number_of_rounds = (num_atoms - 1) / blockDim.x + 1;  // rounds needed to sum over all atoms
+
+    for (int round = 0; round < number_of_rounds; ++round) {
+      int n = tid + round * blockDim.x;
+      if (n < num_atoms) {
+
+        int group = group_ids[n];
+        if (group < 0 || group >= num_groups) return;
+
+        double dx = g_x[n] - g_x_start[size_sum + n];
+        double dy = g_y[n] - g_y_start[size_sum + n];
+        double dz = g_z[n] - g_z_start[size_sum + n];
+
+        if (bid <= correlation_step) {
+          int offset = group * num_correlation_steps + correlation_step - bid;
+          atomicAdd(&g_msd_x[offset], dx * dx);
+          atomicAdd(&g_msd_y[offset], dy * dy);
+          atomicAdd(&g_msd_z[offset], dz * dz);
+        } else {
+          int offset = group * num_correlation_steps + correlation_step + gridDim.x - bid;
+          atomicAdd(&g_msd_x[offset], dx * dx);
```

Comment:
this will not compile for Kepler GPU

Reply:
Nice catch! Fair enough about `atomicAdd`. Do you want me to change that before we merge this?

Reply:
depends on you. At least add that fix for Kepler.